tests/protocol/test_wire.py

```diff
@@ -566,6 +563,176 @@ def test_report_event_large_event(self, patch_send_event, *args):
+        client.save_or_update_goal_state_file(incarnation, xml_text)
+
+        # Assert the file exists and its contents
+        self.assertTrue(os.path.exists(goal_state_file))
+        with open(goal_state_file, "r") as f:
+            contents = f.readlines()
+            self.assertEquals("".join(contents), xml_text)
+
+    def test_save_or_update_goal_state_should_update_existing_goal_state_file(self):
+        incarnation = 42
+        goal_state_file = os.path.join(conf.get_lib_dir(), "GoalState.{0}.xml".format(incarnation))
+        xml_text = WireProtocolData(DATA_FILE).goal_state
+
+        with open(goal_state_file, "w") as f:
+            f.write(xml_text)
+
+        # Assert the file exists and its contents
+        self.assertTrue(os.path.exists(goal_state_file))
+        with open(goal_state_file, "r") as f:
+            contents = f.readlines()
+            self.assertEquals("".join(contents), xml_text)
+
+        # Update the container id
+        new_goal_state = WireProtocolData(DATA_FILE).goal_state.replace("c6d5526c-5ac2-4200-b6e2-56f2b70c5ab2",
+                                                                        "z6d5526c-5ac2-4200-b6e2-56f2b70c5ab2")
+        client = WireClient(wireserver_url)
+        client.save_or_update_goal_state_file(incarnation, new_goal_state)
+
+        # Assert the file exists and its contents
+        self.assertTrue(os.path.exists(goal_state_file))
+        with open(goal_state_file, "r") as f:
+            contents = f.readlines()
+            self.assertEquals("".join(contents), new_goal_state)
+
+    def test_save_or_update_goal_state_should_update_goal_state_and_container_id_when_not_forced(self):
+        incarnation = "1"  # Match the incarnation number from dummy goal state file
+        incarnation_file = os.path.join(conf.get_lib_dir(), INCARNATION_FILE_NAME)
+        with open(incarnation_file, "w") as f:
+            f.write(incarnation)
+
+        xml_text = WireProtocolData(DATA_FILE).goal_state
+        goal_state_file = os.path.join(conf.get_lib_dir(), "GoalState.{0}.xml".format(incarnation))
+
+        with open(goal_state_file, "w") as f:
+            f.write(xml_text)
+
+        client = WireClient(wireserver_url)
+        host = client.get_host_plugin()
+        old_container_id = host.container_id
+
+        # Update the container id
+        new_goal_state = WireProtocolData(DATA_FILE).goal_state.replace("c6d5526c-5ac2-4200-b6e2-56f2b70c5ab2",
+                                                                        "z6d5526c-5ac2-4200-b6e2-56f2b70c5ab2")
+        with patch("azurelinuxagent.common.protocol.wire.WireClient.fetch_config", return_value=new_goal_state):
+            client.update_goal_state(forced=False)
+
+        self.assertNotEqual(old_container_id, host.container_id)
+        self.assertEquals(host.container_id, "z6d5526c-5ac2-4200-b6e2-56f2b70c5ab2")
+
+    def test_send_request_using_appropriate_channel_direct_func_works(self, *args):
```

Comment:
I'd suggest naming this test something like "test_send_request_using_appropriate_channel_should_not_invoke_host_channel_when_direct_channel_succeeds"